# Analyse Exploratoire

### Import des modules

In [ ]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns 

### Analyse Exploratoire
#### importation du dataset
Importer le dataset dans une variable qui sera utilisée a plusieurs reprises durant ce projet

In [ ]:
building_consumption = pd.read_csv("data/raw/2016_Building_Energy_Benchmarking.csv")

Ici c'est pour verifier que le dataset est bien importé et avoir une premiere vue des données

In [ ]:
building_consumption

#### Audit du dataset 

Vue d'ensemble

In [ ]:
# les dimensions (nb lignes , nb colonnes)
building_consumption.shape

In [ ]:
# Les informations générales sur le dataset , permet aussi de voir ou sont les valeurs nulles
# En comparant le nb de lignes et les non-null 
building_consumption.info()

In [ ]:
# Description détaillée de chaque colonne 
building_consumption.describe(include="all")

#### Création du dataset 'Allégé'
L'intérêt est de créer un dataset personnalisé et plus léger pour permettre un meilleure exploration des données

Suppression des lignes de type résidentiel

In [ ]:
#Identification de toutes les lignes de type résidential 
building_consumption['BuildingType'].value_counts()

In [ ]:
#  Liste des types de bâtiments à exclure
multifamily_types = [
    'Multifamily LR (1-4)', 
    'Multifamily MR (5-9)', 
    'Multifamily HR (10+)'
]

# Filtrer pour ne garder que les lignes qui NE SONT PAS dans cette liste
building_consumption_no_residential = building_consumption[~building_consumption['BuildingType'].isin(multifamily_types)]

Verification que le nouveau dataframe est bien réduit en nombre de lignes

In [130]:
building_consumption_no_residential.shape

(1668, 46)

Les lignes de type résidentiel ont été supprimées
Le nombre de ligne total du dataset était de 3776 et est passé a 1668

Suppression des colonnes (variables) qui ne sont pas nécessaire a l'analyse de données et au ML 


<table style="width: 100%; border-collapse: collapse; border: none;">
  <tr>
    <td style="vertical-align: top; width: 32%; padding-right: 15px; border: none;">
      <table border="1" style="width: 100%; border-collapse: collapse;">
        <thead>
          <tr>
            <th style="padding: 8px; text-align: left;">Colonnes</th>
          </tr>
        </thead>
        <tbody>
          <tr><td style="padding: 6px;">YearBuilt</td></tr>
          <tr><td style="padding: 6px;">NumberofBuildings</td></tr>
          <tr><td style="padding: 6px;">NumberofFloors</td></tr>
          <tr><td style="padding: 6px;">PropertyGFATotal</td></tr>
          <tr><td style="padding: 6px;">PropertyGFAParking</td></tr>
          <tr><td style="padding: 6px;">PropertyGFABuilding(s)</td></tr>
          <tr><td style="padding: 6px;">ListOfAllPropertyUseTypes</td></tr>
          <tr><td style="padding: 6px;">SiteEUI(kBtu/sf)</td></tr>
          <tr><td style="padding: 6px;">SiteEnergyUse(kBtu)</td></tr>
          <tr><td style="padding: 6px;">SteamUse(kBtu)</td></tr>
          <tr><td style="padding: 6px;">Electricity(kBtu)</td></tr>
          <tr><td style="padding: 6px;">NaturalGas(kBtu)</td></tr>
          <tr><td style="padding: 6px;">DefaultData</td></tr>
          <tr><td style="padding: 6px;">ComplianceStatus</td></tr>
          <tr><td style="padding: 6px;">TotalGHGEmissions</td></tr>
        </tbody>
      </table>
    </td>
    <td style="vertical-align: top; width: 68%; border: none;">
      <table border="1" style="width: 100%; border-collapse: collapse;">
        <thead>
          <tr>
            <th style="padding: 8px; text-align: left;">À supprimer</th>
            <th style="padding: 8px; text-align: left;">Justification</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td style="padding: 6px;">OSEBuildingID</td>
            <td style="padding: 6px;">L’identifiant unique n'est pas pertinent à garder car nous ne souhaitons pas identifier un batiment en particulier</td>
          </tr>
          <tr>
            <td style="padding: 6px;">DataYear</td>
            <td style="padding: 6px;">Pas utile<br>Toutes les lignes ont la même année : 2016<br><code>building_consumption['DataYear'].value_counts(dropna=False)</code></td>
          </tr>
          <tr>
            <td style="padding: 6px;">BuildingType</td>
            <td style="padding: 6px;">Pas utile à garder pour l'analyse mais nécessaire pour filtrer les habitations pour ne pas les traiter</td>
          </tr>
          <tr>
            <td style="padding: 6px;">PrimaryPropertyType</td>
            <td style="padding: 6px;">Pas nécessaire de les garder car la variable <code>ListOfAllPropertyUseTypes</code> le fait aussi et en mieux</td>
          </tr>
          <tr>
            <td style="padding: 6px;">PropertyName</td>
            <td style="padding: 6px;">Pas nécessaire car pas besoin d'identifier un batiment en particulier</td>
          </tr>
          <tr>
            <td style="padding: 6px;">Address</td>
            <td style="padding: 6px;">Pas besoin de cette information pour l'analyse</td>
          </tr>
          <tr>
            <td style="padding: 6px;">City</td>
            <td style="padding: 6px;">Pas utile<br>Toutes les lignes ont la même année : Seattle<br><code>building_consumption['City'].value_counts(dropna=False)</code></td>
          </tr>
          <tr>
            <td style="padding: 6px;">State</td>
            <td style="padding: 6px;">Pas utile<br>Toutes les lignes ont la même année : WA<br><code>building_consumption['City'].value_counts(dropna=False)</code></td>
          </tr>
          <tr>
            <td style="padding: 6px;">ZipCode</td>
            <td style="padding: 6px;">Pas utile<br>Nous ne cherchons pas a savoir ou le batiment se situe (Seattle est suffisant)</td>
          </tr>
          <tr>
            <td style="padding: 6px;">TaxParcelIdentificationNumber</td>
            <td style="padding: 6px;">Pas utile (tout comme l'identifiant unique)</td>
          </tr>
          <tr>
            <td style="padding: 6px;">CouncilDistrictCode</td>
            <td style="padding: 6px;">Pas utile<br>Nous ne cherchons pas a savoir ou le batiment se situe (Seattle est suffisant)</td>
          </tr>
          <tr>
            <td style="padding: 6px;">Neighborhood</td>
            <td style="padding: 6px;">Pas utile<br>Nous ne cherchons pas a savoir ou le batiment se situe (Seattle est suffisant)</td>
          </tr>
          <tr>
            <td style="padding: 6px;">Latitude</td>
            <td style="padding: 6px;">Pas utile<br>Nous ne cherchons pas a savoir ou le batiment se situe (Seattle est suffisant)</td>
          </tr>
          <tr>
            <td style="padding: 6px;">Longitude</td>
            <td style="padding: 6px;">Pas utile<br>Nous ne cherchons pas a savoir ou le batiment se situe (Seattle est suffisant)</td>
          </tr>
          <tr>
            <td style="padding: 6px;">LargestPropertyUseType</td>
            <td style="padding: 6px;">Les surfaces de chaque type d'utilisation n'est pas nécessaire</td>
          </tr>
          <tr>
            <td style="padding: 6px;">LargestPropertyUseTypeGFA</td>
            <td style="padding: 6px;">Les surfaces de chaque type d'utilisation n'est pas nécessaire</td>
          </tr>
          <tr>
            <td style="padding: 6px;">SecondLargestPropertyUseType</td>
            <td style="padding: 6px;">Les surfaces de chaque type d'utilisation n'est pas nécessaire</td>
          </tr>
          <tr>
            <td style="padding: 6px;">SecondLargestPropertyUseTypeGFA</td>
            <td style="padding: 6px;">Les surfaces de chaque type d'utilisation n'est pas nécessaire</td>
          </tr>
          <tr>
            <td style="padding: 6px;">ThirdLargestPropertyUseType</td>
            <td style="padding: 6px;">Les surfaces de chaque type d'utilisation n'est pas nécessaire</td>
          </tr>
          <tr>
            <td style="padding: 6px;">ThirdLargestPropertyUseTypeGFA</td>
            <td style="padding: 6px;">Les surfaces de chaque type d'utilisation n'est pas nécessaire</td>
          </tr>
          <tr>
            <td style="padding: 6px;">YearsENERGYSTARCertified</td>
            <td style="padding: 6px;">Pas utile</td>
          </tr>
          <tr>
            <td style="padding: 6px;">ENERGYSTARScore</td>
            <td style="padding: 6px;">Pas utile</td>
          </tr>
          <tr>
            <td style="padding: 6px;">SiteEUIWN(kBtu/sf)</td>
            <td style="padding: 6px;">Pas utile<br>La ligne SiteEUI(kBtu/sf) donne un résultat plus pertinent</td>
          </tr>
          <tr>
            <td style="padding: 6px;">SourceEUI(kBtu/sf)</td>
            <td style="padding: 6px;">Pas utile , nous voulons savoir quelle energie consommée et non a la source</td>
          </tr>
          <tr>
            <td style="padding: 6px;">SourceEUIWN(kBtu/sf)</td>
            <td style="padding: 6px;">Pas utile , nous voulons savoir quelle energie consommée et non a la source</td>
          </tr>
          <tr>
            <td style="padding: 6px;">SiteEnergyUseWN(kBtu)</td>
            <td style="padding: 6px;">Pas utile , nous voulons savoir quelle energie consommée et non a la source</td>
          </tr>
          <tr>
            <td style="padding: 6px;">Electricity(kWh)</td>
            <td style="padding: 6px;">Pas utile , la variable Electricity(kBtu) donne la meme information avec la bonne unité de mesure</td>
          </tr>
          <tr>
            <td style="padding: 6px;">NaturalGas(therms)</td>
            <td style="padding: 6px;">Pas utile , la variable NaturalGas(kBtu) donne la meme information avec la bonne unité de mesure</td>
          </tr>
          <tr>
            <td style="padding: 6px;">Comments</td>
            <td style="padding: 6px;">Pas utile car toutes vides<br><code>building_consumption['Comments'].value_counts(dropna=False)</code></td>
          </tr>
          <tr>
            <td style="padding: 6px;">Outlier</td>
            <td style="padding: 6px;">Pas exploitable car presque toutes vides <br><code>building_consumption['Outlier'].value_counts(dropna=False)</code></td>
          </tr>
          <tr>
            <td style="padding: 6px;">GHGEmissionsIntensity</td>
            <td style="padding: 6px;">Pas utile car nous mesurons le total et pas l'intensité des emission de gaz a effet de serre</td>
          </tr>
        </tbody>
      </table>
    </td>
  </tr>
</table>


In [ ]:
#  Création d'un nouveau dataset 'copie' pour ne pas travailler sur l'original
building_consumption_light = building_consumption_no_residential

#  Suppression des colonnes qui ne seront pas nécessaire a l'anayse exploratoire et au Machine Learning
building_consumption_light = building_consumption_light.drop([
                                                        'OSEBuildingID',
                                                        'DataYear',
                                                        'BuildingType',
                                                        'PrimaryPropertyType',
                                                        'PropertyName',
                                                        'Address',
                                                        'City',
                                                        'State',
                                                        'ZipCode',
                                                        'TaxParcelIdentificationNumber',
                                                        'CouncilDistrictCode',
                                                        'Neighborhood',
                                                        'Latitude',
                                                        'Longitude',
                                                        'LargestPropertyUseType',
                                                        'LargestPropertyUseTypeGFA',
                                                        'SecondLargestPropertyUseType',
                                                        'SecondLargestPropertyUseTypeGFA',
                                                        'ThirdLargestPropertyUseType',
                                                        'ThirdLargestPropertyUseTypeGFA',
                                                        'YearsENERGYSTARCertified',
                                                        'ENERGYSTARScore',
                                                        'SiteEUIWN(kBtu/sf)',
                                                        'SourceEUI(kBtu/sf)',
                                                        'SourceEUIWN(kBtu/sf)',
                                                        'SiteEnergyUseWN(kBtu)',
                                                        'Electricity(kWh)',
                                                        'NaturalGas(therms)',
                                                        'Outlier',
                                                        'Comments',
                                                        'GHGEmissionsIntensity'], axis=1)


Verification que les colonnes n'apparaissent pas dans le nouveau dataset 

In [127]:
building_consumption_light.shape

(1668, 16)

Comme nous avons supprimé plusieurs colonnes , le dataSet change de dimensions pour passer de (3376 :46) à (1668 : 15)


On peut créer une copie du dataset original (sans les informations inutiles pour ce projet)

In [ ]:
# Création du CSV avec le dataset allégé
building_consumption_light.to_csv('data/Processed/building_consumption_light.csv')

#### Detection des nulls sur la distribution allégée
En faisant cette manipulation on constate que les variables Cibles SiteEnergyUseWN(kBtu) et GHGEmissionsIntensity ont des nulls

In [149]:
df_nulls_light = building_consumption_light.isna().sum().reset_index()
df_nulls_light.columns = ["variable", "nb_nulls"]
df_nulls_light['Frequence'] = df_nulls_light['nb_nulls'] / len(building_consumption_light)
df_nulls_light = df_nulls_light.sort_values('Frequence', ascending=False)
df_nulls_light


,variable,nb_nulls,Frequence
7,SiteEUI(kBtu/sf),3,0.001799
1,NumberofBuildings,2,0.001199
6,ListOfAllPropertyUseTypes,2,0.001199
8,SiteEnergyUse(kBtu),2,0.001199
9,SteamUse(kBtu),2,0.001199
10,Electricity(kBtu),2,0.001199
11,NaturalGas(kBtu),2,0.001199
14,TotalGHGEmissions,2,0.001199
15,FE1_Is_multi_type,2,0.001199
0,YearBuilt,0,0.000000


CONSEIL = Ajouter la fréquence des nulls par colonne

imputer les manquantes (moyenne , médiane)

#### Statistiques descriptives (sur le dataset 'allégé')

##### Les Correlations 
    SiteEnergyUse(kBtu)   =   SteamUse(kBtu) + Electricity(kBtu) + NaturalGas(kBtu)
    PropertyGFATotal   =   PropertyGFAParking + PropertyGFABuilding(s)

tracer l'obeservation graphique 
Si quantitatives >> nuage de point 
Si qualitatives et quantitatives >> boxplots anova ou kurskal wallis
si qualitatives >> tableau de conteingence (tableau croisé) Khi 2 ou V cramer 


Metriques : Pearson, spearman, 

#### Valeurs aberrantes


In [ ]:
# normaliser les données des 4 variables 'energie et des 3 variables 'surfaces'pour avoir un résultat pertinent sur le pairplot

# sns.pairplot(building_consumption_light[['SiteEnergyUse(kBtu)', 'SteamUse(kBtu)', 'Electricity(kBtu)', 'NaturalGas(kBtu)']])
# sns.pairplot(building_consumption_light[['PropertyGFATotal', 'PropertyGFAParking', 'PropertyGFABuilding(s)']])

VISUALISATION DES DONNEES : somme des buildings / annéesConstruction

In [ ]:

comptes = building_consumption_light['YearBuilt'].value_counts().sort_index()

plt.bar(comptes.index, comptes.values)
plt.xlabel("Année de construction")
plt.ylabel("Nombre de bâtiments")
plt.show()

Avec ce diagramme en barres , je souhaitais verifier que les batiments de la ville ont été construits de maniere espacés et réguliere. Ce qui est le cas
Pour l'année 1900, il peut y avoir une date par défaut ce qui explique cette valeur elevée.

VISUALISATION DES DONNEES : Batiments Simple-type / Batiments Multi-Type 

In [ ]:
# preparation des données pour les afficher en diagramme
# Création d'une nouvelle colonne FE1_Is_multi_type 
building_consumption_light['FE1_Is_multi_type'] = (
    building_consumption_light['ListOfAllPropertyUseTypes']
    .dropna()
    .str.split(',')
    .apply(len)
)

# Transformation des données en 0 et 1 pour faciliter le ML 
building_consumption_light['FE1_Is_multi_type'] = building_consumption_light['FE1_Is_multi_type'].apply(
    lambda x: 0 if x == 1 else (1 if x >= 2 else x)
)

# Change le type des données de FLOAT en INT (pour eviter les virgules)
building_consumption_light['FE1_Is_multi_type'] = building_consumption_light['FE1_Is_multi_type'].astype('Int64')


In [ ]:
# Diagramme en barres : simple usage / multi usage
comptes_types= building_consumption_light['FE1_Is_multi_type'].value_counts()

plt.bar(comptes_types.index, comptes_types.values)
plt.xticks(
    ticks=[0, 1], 
    labels=['Mono Usage', 'Multi Usage'], 
    rotation=0
)
plt.grid(axis='y')
plt.title("Proportion des batiments mono et multi usage")



Avec ce diagramme , nous pouvons constater qu'il y a une proportion presque equivalente des batiments mono-usage et multi-usage

In [ ]:
# GRAPH  : surface / consomation d'energie

In [ ]:
# GRAPH  : surface / total emissions CO2

#### TERMINER L'ANALYSE EXPLORATOIRE 

A réaliser : 
- Une analyse descriptive des données, y compris une explication du sens des colonnes gardées, des arguments derrière la suppression de lignes ou de colonnes, des statistiques descriptives et des visualisations pertinentes.

Qelques pistes d'analyse : 

* Identifier les colonnes avec une majorité de valeurs manquantes ou constantes en utilisant la méthode value_counts() de Pandas
* Mettre en evidence les différences entre les immeubles mono et multi-usages
* Utiliser des pairplots et des boxplots pour faire ressortir les outliers ou des batiments avec des valeurs peu cohérentes d'un point de vue métier 

Pour vous inspirer, ou comprendre l'esprit recherché dans une analyse exploratoire, vous pouvez consulter ce notebook en ligne : https://www.kaggle.com/code/pmarcelino/comprehensive-data-exploration-with-python. Il ne s'agit pas d'un modèle à suivre à la lettre ni d'un template d'analyses attendues pour ce projet. 

# Modélisation 

### Import des modules 

In [ ]:
#Selection
from sklearn.model_selection import (
    train_test_split,
    GridSearchCV, 
    cross_validate,
)
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error 
from sklearn.inspection import permutation_importance

#Preprocess
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import LabelEncoder, OneHotEncoder, StandardScaler

#Modèles
from sklearn.dummy import DummyRegressor
from sklearn.linear_model import LinearRegression
from sklearn.svm import SVR
from sklearn.ensemble import RandomForestRegressor


### Feature Engineering

A réaliser : Enrichir le jeu de données actuel avec de nouvelles features issues de celles existantes. 

En règle générale : On utilise la méthode .apply() de Pandas pour créer une nouvelle colonne à partir d'une colonne existante. N'hésitez pas à regarder les exemples dans les chapitres de cours donnés en ressource

In [ ]:
# CODE FEATURE ENGINEERING

### Préparation des features pour la modélisation

A réaliser :
* Si ce n'est pas déjà fait, supprimer toutes les colonnes peu pertinentes pour la modélisation.
* Tracer la distribution de la cible pour vous familiariser avec l'ordre de grandeur. En cas d'outliers, mettez en place une démarche pour les supprimer.
* Débarrassez-vous des features redondantes en utilisant une matrice de corrélation de Pearson. Pour cela, utiisez la méthode corr() de Pandas, couplé d'un graphique Heatmap de la librairie Seaborn 
* Réalisez différents graphiques pour comprendre le lien entre vos features et la target (boxplots, scatterplots, pairplot si votre nombre de features numériques n'est pas très élevé).
*  Séparez votre jeu de données en un Pandas DataFrame X (ensemble de feautures) et Pandas Series y (votre target).
* Si vous avez des features catégorielles, il faut les encoder pour que votre modèle fonctionne. Les deux méthodes d'encodage à connaitre sont le OneHotEncoder et le LabelEncoder

In [ ]:
# CODE PREPARATION DES FEATURES

### Comparaison de différents modèles supervisés

A réaliser :
* Pour chaque algorithme que vous allez tester, vous devez :
    * Réaliser au préalable une séparation en jeu d'apprentissage et jeu de test via une validation croisée.
    * Si les features quantitatives que vous souhaitez utiliser ont des ordres de grandeur très différents les uns des autres, et que vous utilisez un algorithme de regression qui est sensible à cette différence, alors il faut réaliser un scaling (normalisation) de la donnée au préalable.
    * Entrainer le modèle sur le jeu de Train
    * Prédire la cible sur la donnée de test (nous appelons cette étape, l'inférence).
    * Calculer les métriques de performance R2, MAE et RMSE sur le jeu de train et de test.
    * Interpréter les résultats pour juger de la fiabilité de l'algorithme.
* Vous pouvez choisir par exemple de tester un modèle linéaire, un modèle à base d'arbres et un modèle de type SVM
* Déterminer le modèle le plus performant parmi ceux testés.

In [ ]:
# CODE COMPARAISON DES MODELES

### Optimisation et interprétation du modèle

A réaliser :
* Reprennez le meilleur algorithme que vous avez sécurisé via l'étape précédente, et réalisez une GridSearch de petite taille sur au moins 3 hyperparamètres.
* Si le meilleur modèle fait partie de la famille des modèles à arbres (RandomForest, GradientBoosting) alors utilisez la fonctionnalité feature importance pour identifier les features les plus impactantes sur la performance du modèle. Sinon, utilisez la méthode Permutation Importance de sklearn.

In [ ]:
# CODE OPTIMISATION ET INTERPRETATION DU MODELE